[d3d12 wgl] Upgrade to `windows 0.62` crates

MarijnS95 opened the pull request:
**Connections**

https://github.com/microsoft/windows-rs/releases/tag/0.61.0
https://github.com/microsoft/windows-rs/releases/tag/63
https://github.com/microsoft/windows-rs/releases/tag/69

**Description**


Uses the latest `windows 0.62` and `windows-core 0.62` releases, including some minor code improvements for us.  The MSRV has been bumped to `1.74`, but `wgpu` is already on `1.76` anyway.

**Testing**
```sh
WGPU_BACKEND=d3d12 cargo r -p wgpu-examples hello_triangle
```

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [ ] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.
- [ ] Replaced crate patches with actual releases.

MarijnS95 commented on wgpu-hal/src/gles/wgl.rs:
```diff
@@ -127,7 +127,7 @@ impl WglContext {
             return Ok(());
         }
-        unsafe { OpenGL::wglMakeCurrent(None, None) }
+        unsafe { OpenGL::wglMakeCurrent(Default::default(), Default::default()) }
```
@kennykerr can you help us here? I did not do my due diligence this time to search through `windows-rs` and `win32metadata` for relevant changes, but while most handles have been made optional (see other `Some()` wrapping changes in this PR), quite strangely these parameters are no longer optional in `wglMakeCurrent()` where they have a specific use-case to uncurrent the context.

Is this change intended or accidental? We can clearly get around it by passing a defaulted `HDC`/`HGLRC`, but it seems strange when compared to other functions where `None` can be passed.

kennykerr replied:
The code generator now faithfully honors the metadata's `[Optional]` attribute - in the case of `wglMakeCurrent` the handle parameters are not optional whereas in the case of `CreateSwapChainForCompositionSurfaceHandle` the handle parameter is optional.

MarijnS95 replied:
Thanks @kennykerr. So this'll take another trip through `win32metadata` to annotate correctly.

cwfitzgerald commented:
Just to set expectations here, we likely can't merge this for a bit, as firefox is currently on 0.58 and migration will be a bit of a thing. @ErichDonGubler will take charge of this.

I don't expect there to be a ton of problems with this PR just hanging out for a while. I'm going to mark this as draft as we can't merge it, but @Wumpf will review it still.

MarijnS95 commented:
@cwfitzgerald that's all good, I don't expect this to be merged until the upgrade has fully trickled through the ecosystem. Just making sure we have a PR open with all the necessary changes that I couldn't make in the original `windows-rs` migration due to missing or incorrect upstream annotations.

ErichDonGubler commented:
There are a few things that I'm already aware need to happen in `mozilla-central` before this can land:

- [ ] Dependencies on `windows`
	- [ ] Bump `gpu-allocator` after a version releases that consumes https://github.com/Traverse-Research/gpu-allocator/pull/258.
		- We might need this backported to 0.27.0, since mainline upstream now uses `objc2-metal`, which we're blocked on moving to because of #5641 (CC @Jasper-Bekkers, @MarijnS95).
	- [x] [`mtu`](https://docs.rs/mtu) migrated from `windows-{bindgen,core}` and `windows-targets` to `windows` `>=0.58.0,<0.60.0`: https://github.com/mozilla/mtu/pull/65
- [ ] Dependencies on `windows-sys`
	- [x] `mio`: https://github.com/tokio-rs/mio/pull/1857
	- [ ] Coordinate upgrades with folks working on [`neqo`](https://github.com/mozilla/neqo): https://github.com/mozilla/neqo/pull/2662
	- [x] `audio_thread_priority` needs migration: https://github.com/mozilla/audio_thread_priority/pull/30
	- [x] ~~`tempfile` can have its version range extended: https://github.com/Stebalien/tempfile/pull/319~~ Actually already works, but it has a nitpick issue with its range specification.
	- [x] `socket2` can have its version range extended:
		- https://github.com/rust-lang/socket2/pull/545
		- https://github.com/rust-lang/socket2/pull/579
	- [x] `tokio` needs migration: https://github.com/tokio-rs/tokio/pull/7117
		- [x] `is-terminal` can have its version range extended: https://github.com/sunfishcode/is-terminal/pull/43
- [ ] Direct dependencies on `windows-targets`
	- [x] `libloading` can widen its accepted ranges of `windows-targets` to include 0.53.\*: https://github.com/nagisa/rust_libloading/pull/169
	- [ ] `backtrace`: consume a new `libloading` and update workspace: https://github.com/rust-lang/backtrace-rs/pull/694
	- [ ] `parking_lot_core`: consume a new `backtrace` and update workspace: https://github.com/Amanieu/parking_lot/pull/458

larseggert commented:
Regarding `neqo` and `mtu`, we can bump whenever you're ready - just let us know. (CC @mxinden for info).

mxinden commented:
> Regarding `neqo` and `mtu`, we can bump whenever you're ready - just let us know. (CC @mxinden for info).

Under the assumption that `windows` `v0.59` does not introduce a breaking change for `neqo` or `mtu`, we can also make `neqo` and `mtu` compatible to both `windows` `v0.58` and `v0.59` by configuring a range:
``` toml
windows-sys = { version = ">=0.58, <=0.59"
```

See https://github.com/quinn-rs/quinn/pull/2021/ as an example from one of our upstream dependencies.

@ErichDonGubler let me know what works best for you.

ErichDonGubler commented:
@mxinden: If a range dependency works for `neqo` and `mtu`, then _wonderful_, let's please do that. 😀 Did you want me to file PRs?

ETA: Filed! See the comment (https://github.com/gfx-rs/wgpu/pull/6876#issuecomment-2581684258) for details.

Wumpf reviewed (approved):
fwiw code lgtm and took it on a spin running examples locally

ErichDonGubler commented:
https://github.com/mozilla/mtu/pull/65 has landed! 🙌🏻

ErichDonGubler commented:
I've updated my [dependency index comment](https://github.com/gfx-rs/wgpu/pull/6876#issuecomment-2581684258) to be organized by direct dependency. I've filed the PRs that I'm aware are necessary in the larger Crates.io ecosystem, and am now ready (but not yet trying in earnest) to try compiling `mozilla-central` with the updated deps.

MarijnS95 commented:
>   * [ ]  Bump `gpu-allocator` after a version releases that consumes [cargo: Bump `windows` crate range to `0.53-0.59` Traverse-Research/gpu-allocator#258](https://github.com/Traverse-Research/gpu-allocator/pull/258).
>     
>     * We might need this backported to 0.27.0, since mainline upstream now uses `objc2-metal`, which we're blocked on moving to because of [[metal] Use `objc2-metal` #5641](https://github.com/gfx-rs/wgpu/pull/5641) (CC @Jasper-Bekkers, @MarijnS95).

@ErichDonGubler I think we can arrange this pretty quickly, let us know when the time is there (and you haven't yet upgraded to `objc2-metal` by then).

Wumpf commented:
@ErichDonGubler any update on the Mozilla side? looks like a lot of the PRs you [referenced](https://github.com/gfx-rs/wgpu/pull/6876#issuecomment-2581684258) are landed by now

MarijnS95 commented:
Sorry I completely dropped the ball on this and there are quite a few new versions already. I'll get back to the second bump ASAP, as long as Mozilla is okay with the next bumps?

ErichDonGubler commented:
@MarijnS95: The way to keep this tractable is almost certainly to keep uograde steps incremental, so don't feel like you need to throw away efforts to target 0.59 without a strong concrete blocker.

ErichDonGubler commented:
@Wumpf: Some PRs have landed, and some have released, but many of them have not been exposed in a Crates.io release yet. I'm checking those boxes as releases become available on Crates.io, not when the PRs land.

MarijnS95 commented:
@ErichDonGubler It is expected that the upgrade to `0.61`  is incremental on top of `0.59`, nothing to be thrown away. But this PR is definitely in need of a rebase and conflict-resolve.

MarijnS95 commented:
@ErichDonGubler you can cross off `mio` which released `1.0.4` with the bump just 6 hours ago (`tokio` also released, but the bump wasn't merged yet).

MarijnS95 commented on Cargo.toml:
```diff
@@ -249,6 +249,11 @@ ndk-sys = "0.6"
 
+# https://github.com/Traverse-Research/gpu-allocator/pull/258
+gpu-allocator = { git = "https://github.com/Traverse-Research/gpu-allocator", rev = "955b13a" }
+# https://github.com/Xudong-Huang/generator-rs/pull/72
+generator = { git = "https://github.com/MarijnS95/generator-rs", rev = "e4afb28" }
```
Looks like @Xudong-Huang [kept ignoring](https://github.com/Xudong-Huang/generator-rs/pull/72#issuecomment-2645924543) my PR and silently jumped straight to `0.60` in https://github.com/Xudong-Huang/generator-rs/commit/250f96ae7ba5e916ed2a83d3c216568f5f810e24.

MarijnS95 replied:
Without a range dependency, which would have been possible, `wgpu` can't deduplicate on `windows-rs`. A later `clippy` commit bumped to `0.61`: https://github.com/Xudong-Huang/generator-rs/commit/643c7138fee58deac2a298c25b57a4f621d451da.

I've updated my PR to change to a range.

MarijnS95 commented:
I'm a little confused by the latest CI error https://github.com/gfx-rs/wgpu/actions/runs/16555966628/job/46817429146?pr=6876:

```
error: failed to select a version for `libc`.
    ... required by package `socket2 v0.6.0`
    ... which satisfies dependency `socket2 = "^0.6.0"` of package `tokio v1.47.0`
    ... which satisfies dependency `tokio = "^1.47"` of package `cts_runner v26.0.0 (/home/runner/work/wgpu/wgpu/cts_runner)`
versions that meet the requirements `^0.2.172` are: 0.2.174, 0.2.173, 0.2.172

all possible versions conflict with previously selected packages.

  previously selected package `libc v0.2.168`
    ... which satisfies dependency `libc = "^0.2.168"` of package `wgpu-hal v26.0.0 (/home/runner/work/wgpu/wgpu/wgpu-hal)`
    ... which satisfies path dependency `wgpu-hal` of package `wgpu v26.0.0 (/home/runner/work/wgpu/wgpu/wgpu)`
    ... which satisfies path dependency `wgpu` of package `wgpu-benchmark v26.0.0 (/home/runner/work/wgpu/wgpu/benches)`

failed to select a version for `libc` which could resolve this conflict
```

This PR only bumps `socket2` from `0.5.10` to `0.6` via `tokio 1.47`. [That `0.5.10` already had a `libc 0.2.171` bound](https://crates.io/crates/socket2/0.5.10/dependencies), which should similarly be incompatible with that "previously selected `libc 0.2.168`" for `wgpu-hal`. However, it seems then the resolver understood it's an open-ended dependency and `wgpu-hal` can trivially use `libc 0.2.172`?

The solution is probably simple, just bump our dep to `libc = "0.2.172"`, I just feel like understanding why it didn't need to be `libc = "0.2.171"` on `trunk` already.

cwfitzgerald commented:
This is a really weird error that you hit with `-Zdirect-minimal-versions`. When you have a dependency that has a higher requirement than you do, instead of resolving it to that higher version, it throws an error as it's trying to lock your immediate versions to exactly the versions you need. This may end up being a unnecessary requirement, but would require downgrading a transitive dependency, which `-Zdirect-minimal-versions` won't do.

MarijnS95 commented:
Never mind, I was mistaken. While I said that `socket2 0.5.10` requires `libc 0.2.171` already, `tokio 1.64.1` only requires `socket2 0.5.5` which requires `libc 0.2.149`, so the requirement is satisfied when `-Zdirect-minimal-versions` sets it at `libc 0.2.168` for `wgpu-hal`.

cwfitzgerald commented:
@ErichDonGubler how are we doing here, mozilla side?

ErichDonGubler commented:
@cwfitzgerald: The [dependency status index comment](https://github.com/gfx-rs/wgpu/pull/6876#issuecomment-2581684258) is up-to-date, as far as I know. Some upstream PRs have stalled, and now some of the deps. have even started to move on to later `windows` versions. There's also some complication with some upstream movement to `windows-link`, instead of `windows-sys`.

I think we might benefit from retrying with the latest `windows` crate ecosystem, rather than trying to continue with 0.59.

MarijnS95 committed:
[d3d12 wgl] Upgrade to windows 0.62 crates

https://github.com/microsoft/windows-rs/releases/tag/69

MarijnS95 commented:
> I think we might benefit from retrying with the latest `windows` crate ecosystem, rather than trying to continue with 0.59.

@ErichDonGubler with that, are you referring to [`0.61`](https://github.com/microsoft/windows-rs/releases/tag/63) from https://github.com/gfx-rs/wgpu/pull/8013 or the latest [`0.62` release](https://github.com/microsoft/windows-rs/releases/tag/69) which barely misses out on range compatibility due to https://github.com/gfx-rs/wgpu/pull/8013#issuecomment-3271792993?

MarijnS95 committed:
[d3d12 wgl] Upgrade to `windows 0.59` crates

https://github.com/microsoft/windows-rs/releases/tag/0.61.0

The latest `windows 0.59` and `windows-core 0.59` crates were
just released (strangely tagged `0.61`), including some minor code
improvements for us.  The MSRV has been bumped to `1.74`, but `wgpu` is
already on `1.76` anyway.

MarijnS95 commented:
We dropped `gpu-allocator 0.28` :)

ErichDonGubler commented:
@MarijnS95:

> > I think we might benefit from retrying with the latest `windows` crate ecosystem, rather than trying to continue with 0.59.
> 
> @ErichDonGubler with that, are you referring to [`0.61`](https://github.com/microsoft/windows-rs/releases/tag/63) from #8013 or the latest [`0.62` release](https://github.com/microsoft/windows-rs/releases/tag/69) which barely misses out on range compatibility due to [#8013 (comment)](https://github.com/gfx-rs/wgpu/pull/8013#issuecomment-3271792993)?

I'm referring to 0.62, sorry for any confusion!

Much of the ecosystem has caught up with 0.62, with some refusing to use range dependencies. I've also confirmed that we can build it in Firefox, done all the auditing work for new dependencies. The main supply chain task is some upstreaming before review that can be done incrementally per-dependency (see [Bug 1991226 - Upgrade/audit most Rust deps. for WGPU revendor and `windows` 0.62, part 1](https://bugzilla.mozilla.org/show_bug.cgi?id=1991226)). 

I'll let you judge for yourself, though. I've put a PR for 0.62 up&mdash;based on work here&mdash;for your perusal: <https://github.com/gfx-rs/wgpu/pull/8281>

MarijnS95 commented:
> Much of the ecosystem has caught up with 0.62, with some refusing to use range dependencies. 

As already highlighted in https://github.com/gfx-rs/wgpu/pull/8013#issuecomment-3271792993 **three whole weeks ago** `wgpu` itself cannot take a range dependency because of the `Error::from_win32()` -> `Error::from_thread()` rename.

> I'll let you judge for yourself, though. I've put a PR for 0.62 up—based on work here—for your perusal: https://github.com/gfx-rs/wgpu/pull/8281

I thought I had also already shared https://github.com/MarijnS95/wgpu/compare/windows-69 with you those three weeks ago, and was merely waiting for confirmation to push it into this PR or #8013.

MarijnS95 commented:
Note that all these force-pushes to contributor PRs -in particular after such a long time of radio silence- seem a little disingenuous: I cannot even read the CI failure and `git rebase --update-refs` to amend the fixes into multiple PRs at once, before having `git push --force-with-lease` tell me that **my** remote branch changed since checkout? Would be awesome if you could ask next time whether help is needed 🙏

ErichDonGubler commented:
> As already highlighted in [#8013 (comment)](https://github.com/gfx-rs/wgpu/pull/8013#issuecomment-3271792993) **three whole weeks ago** `wgpu` itself cannot take a range dependency because of the `Error::from_win32()` -> `Error::from_thread()` rename.

Sure, but that only means WGPU specifically suffers from being an all-or-nothing upgrade WRT this `windows` upgrade. It's still beneficial for WGPU's dependencies to be incrementally upgradeable, because incremental upgrades at Mozilla have a significantly lower risk of getting stuck. If you examine bug 1991226, that's exactly what I'm doing.

When you submitted that comment three weeks ago, this was, unfortunately, not something I could justify looking at (and I'm still not sure that I can drive this to completion, based on Mozilla's current priorities). I'm trying to split a bit of attention to take a look at this _now_, though, because I think it's important for Mozilla to have concrete reasons for either letting things through or being blocked, rather than "I don't got bandwidth indefinitely, soz."

ErichDonGubler commented:
> Note that all these force-pushes to contributor PRs -in particular after such a long time of radio silence- seem a little disingenuous: I cannot even read the CI failure and `git rebase --update-refs` to amend the fixes into multiple PRs at once, before having `git push --force-with-lease` tell me that **my** remote branch changed since checkout? Would be awesome if you could ask next time whether help is needed 🙏

Ah, sorry, I'm not trying to create obstacles for you! I'm happy to switch to appending `fixup!`s, or giving you diffs via PRs. Force-pushing generally works very well with less dedicated contributions, but you've been quite dedicated. ❤️ Happy to accommodate a better workflow here!

ErichDonGubler commented:
> I thought I had also already shared https://github.com/MarijnS95/wgpu/compare/windows-69 with you those three weeks ago, and was merely waiting for confirmation to push it into this PR or https://github.com/gfx-rs/wgpu/pull/8013.

I don't remember this being shared, but I think it's safe to say we've reached consensus for trying to target `windows` 0.62. I'm fine dismissing #8281, if you intend to reconcile it with #8013.

MarijnS95 commented on .deny.toml:
```diff
@@ -76,6 +74,3 @@ allow-git = [
 unknown-git = "deny"
 required-git-spec = "rev"
-
-[sources.allow-org]
-github = ["gfx-rs"]
```
I have yet to revert this back out. I added `command-arguments: -Dwarnings` to CI to catch some warnings that otherwise went unnoticed, but that includes unmatched `allow-org` when there are (temporarily) no patches from this org...

MarijnS95 commented:
Sure thing, having downstream dependencies commit to ranges (when possible) would be awesome and I've been trying to sneak them into crates whenever possible. Thus happen to contribute to quite a few too many, and it usually means preempting the maintainer or other contributors who are oblivious to this behavior (and getting the range extended backwards has thus far proven harder).

Hadn't looked at the Bugzilla tracker but it looks like a lot of crates to repeatedly vet; any way that could be reduced by less aggressively bumping patch versions where not necessary for e.g. this Windows upgrade?

Glad to see that you're at least trying to get something going here despite not finding time/priority. Apologies for the hasty response. Makes me fear the ever-increasing stack of PRs, issues and (mostly) complaints on popular repos that I'm supposed to be maintaining, ... 🤷

I'll happily keep #8013 up to date and rebased on this PR, and leave it to you whether to temporarily jump to `0.59` via this PR or go straight to `0.62`. Since most context lives here, I can merge it into this PR too.

ErichDonGubler commented:
I'd like to address this tension specifically, since I consider cultural gardening an important exercise:

> Note that all these force-pushes to contributor PRs -in particular after such a long time of radio silence- seem a little disingenuous: ...

As I'm sure you know, being radio-silent generally isn't a problem for one-off contributions. This case is already different; IMO, you've definitely earned some negotiating power on how you'd like to communicate, given how much (welcome) work you've been doing. Thoughts? I'm guessing you would have preferred that we told you specifically that we knew we didn't have immediate bandwidth for `windows` upgrades. Alternatively, we don't mind getting pinged in issues like this if and when we do fail to communicate, though responsiveness can vary between Mozillians and from week to week. Sometimes, we have week-or-longer interruptions (like the recent WebGPU F2F, or Mozilla all-company meetings).

☝🏻 CC @gfx-rs/wgpu.

ErichDonGubler commented:
> Hadn't looked at the Bugzilla tracker but it looks like a lot of crates to repeatedly vet; any way that could be reduced by less aggressively bumping patch versions where not necessary for e.g. this Windows upgrade?

We tend not to repeatedly vet these crates, with the recent move in wgpu from Dependabot to Renovate. It was _awful_ before that point.

Those specific dependencies are getting updated in that patches series because IIRC some other crate required a new minimum version on the road to getting upgraded to support `windows` 0.62. I don't remember specifics for these, though…there were many. 😅

ErichDonGubler commented:
RE: the set of PRs to keep open: I think consolidating everything into this PR would make things significantly easier to follow for everyone. We don't actually intend to land things incrementally, so multiple PRs sound only like additional complication.

MarijnS95 commented:
> As I'm sure you know, being radio-silent generally isn't a problem for one-off contributions. This case is already different; IMO, you've definitely earned some negotiating power on how you'd like to communicate, given how much (welcome) work you've been doing. Thoughts? I'm guessing you would have preferred that we told you specifically that we knew we didn't have immediate bandwidth for `windows` upgrades.

This was mostly in respect to first being on one end of the spectrum (no response for weeks) to then jumping all the way to the other with force-pushes and respin PRs (something I tend to use sparingly when a contributor hasn't responded for weeks and I really want to get their change in after some minor CI/review fixups).

---

> We tend not to repeatedly vet these crates, with the recent move in wgpu from Dependabot to Renovate. It was _awful_ before that point.

Does that mean you're now skipping certain patch releases or just postponing the Renovate PRs until a vet is possible/completed?

> Those specific dependencies are getting updated in that patches series because IIRC some other crate required a new minimum version on the road to getting upgraded to support `windows 0.62. I don't remember specifics for these, though...there were many. 😅

Ah, yup, https://github.com/gfx-rs/wgpu/pull/8013/files#diff-13ee4b2252c9e516a0547f2891aa2105c3ca71c6d7a1e682c69be97998dfc87e hits quite a lot of transitive bumps. Wish those were (mostly) unnecessary 😓

ErichDonGubler commented:
> This was mostly in respect to first being on one end of the spectrum (no response for weeks) to then jumping all the way to the other with force-pushes and respin PRs (something I tend to use sparingly when a contributor hasn't responded for weeks and I really want to get their change in after some minor CI/review fixups).

Ah, that wasn't intended as a respin, sorry! I had pushed branches and a temp PR already so I had a stable point of reference for making sure things were viable in Firefox. I had no illusions that I'd get that through without working with you.

The force-pushes were mostly so I could keep 0.62 changes on top of a rebased set of your patches. I promise I didn't change anything substantially! But then again...force-pushes on others' stuff still usually deserves more explanation than a `Compare` link, so point taken. Sorry about that! 💦 

ETA: Oh, yeah, I had also force-pushed to take care of a TOML formatting failure.

ErichDonGubler commented:
> Does that mean you're now skipping certain patch releases or just postponing the Renovate PRs until a vet is possible/completed?

Actually, we can have our cake and eat it, too. Renovate actually implements bumping `Cargo.lock` without forcing a bump in `Cargo.toml`, so we get benefit from both upgrading and not upgrading. Combined with our MSRV/dep.-minimization checks in CI, it's actually quite nice!

MarijnS95 commented:
I've dropped all transitive dependency upgrades now, the revert seems to mostly contain `windows-sys` related bumps that should be consumed via Renovate instead.

Some `windows` crates had a patch-release from https://github.com/microsoft/windows-rs/releases/70, we can incorporate those as long as they don't affect transitive dependencies.

ErichDonGubler commented:
We'll also need <https://github.com/microsoft/windows-rs/pull/3785>; having case-insensitive collisions b/w `README.md` and `readme.md` tripped up initial integration. That seems on track for another patch release, tho!

MarijnS95 commented:
Oh noes, cased files on a case-insensitive filesystem. /me shudders in `include` vs `Include` used in the Vulkan SDK, I have yet to learn how to best set that up for a Windows cross-compile from a Linux machine (for which no Windows-targeted SDK tarball exists...).

ErichDonGubler commented:
I'm making a new index comment here for the set of dependencies that need an upgrade before Mozilla can consume this. Fortunately, it's a rather small list!

- [x] https://github.com/mozilla/memtest/pull/18
  - CC @gabrielesvelto
- [x] https://github.com/Boddlnagg/midir/pull/178
- [x] https://github.com/mozilla/neqo/pull/3091
  - ☝🏻 CC @larseggert

MarijnS95 committed:
[d3d12 wgl] Upgrade to `windows 0.61` crates

https://github.com/microsoft/windows-rs/releases/tag/63

ErichDonGubler commented:
I'm going to merge this, because the last straggler for upgrade in Firefox (CC @larseggert) has an approved PR, and it's fairly easy to backport that PR to whatever revision base that `neqo` needs in Firefox. Things are looking surprisingly not painful! 👏🏻😭

ErichDonGubler commented:
On behalf of the entire wgpu maintainership in a meeting just now, I'd like to thank @MarijnS95 for the tireless work in this PR to keep things coordinated, and patience with my (sometimes conflicting) efforts to shepherd this forward. It's not easy to move forward an entire software ecosystem, but you've done just that. ❤️

ErichDonGubler commented:
Notes from today's maintainers' meeting:

> - [#6876](https://github.com/gfx-rs/wgpu/pull/6876) [d3d12 wgl] Upgrade to windows 0.62 crates
>   - CF: Are we good to merge this, then?
>   - EG: Yep. Merged!
>   - Consensus: General noises of excitement, relief, and catharsis
